tfe/data_source_oauth_client.go

```diff
@@ -36,16 +80,95 @@ func dataSourceTFEOAuthClientRead(d *schema.ResourceData, meta interface{}) erro
 	ctx := context.TODO()
 	tfeClient := meta.(*tfe.Client)
 
-	ocID := d.Get("oauth_client_id").(string)
+	var oc *tfe.OAuthClient
+	var err error
+
+	if v, ok := d.GetOk("oauth_client_id"); ok {
+		oc, err = tfeClient.OAuthClients.Read(ctx, v.(string))
+		if err != nil {
+			return fmt.Errorf("Error retrieving OAuth client: %w", err)
+		}
+	} else {
+		// search by name or service provider within a specific organization instead
+		organization := d.Get("organization").(string)
+
+		var name string
+		var serviceProvider tfe.ServiceProviderType
+		vName, ok := d.GetOk("name")
+		if ok {
+			name = vName.(string)
+		}
+		vServiceProvider, ok := d.GetOk("service_provider")
+		if ok {
+			serviceProvider = tfe.ServiceProviderType(vServiceProvider.(string))
+		}
 
-	oc, err := tfeClient.OAuthClients.Read(ctx, ocID)
-	if err != nil {
-		return fmt.Errorf("Error retrieving OAuth client: %w", err)
+		// Paginate through all OAuthClients in the organization; if multiple pages
+		// of results are returned by the API, use the options variable to increment
+		// the page number until all results have been retrieved.
+		//
+		// Within the pagination loop, loop again through each result on each page.
+		// If 'name' was set, then match against the 'Name' field. If 'service_provider'
+		// was set, then match against the 'ServiceProvider' field. If both are set,
+		// then both must match. All matches are added to the ocMatches slice.
+		//
+		// At the end of the loop, if zero or more than one matches were found, an
+		// error is returned. Otherwise, only one match was found, and execution
+		// proceeds beyond the loop to build the resource object.
+		//
+		var ocMatches []*tfe.OAuthClient
+		options := &tfe.OAuthClientListOptions{}
+		for {
+			ocList, err := tfeClient.OAuthClients.List(ctx, organization, options)
+			if err != nil {
+				return fmt.Errorf("Error retrieving OAuth Clients: %w", err)
+			}
+
+			for _, item := range ocList.Items {
+				if name != "" && serviceProvider != "" {
+					if item.Name != nil && *item.Name == name && item.ServiceProvider == serviceProvider {
+						ocMatches = append(ocMatches, item)
+					}
+				} else if name != "" {
+					if item.Name != nil && *item.Name == name {
+						ocMatches = append(ocMatches, item)
+					}
+				} else if serviceProvider != "" {
+					if item.ServiceProvider == serviceProvider {
+						ocMatches = append(ocMatches, item)
+					}
+				}
+			}
+
+			// Exit the loop when we've seen all pages.
+			if ocList.CurrentPage >= ocList.TotalPages {
+				break
+			}
+
+			// Update the page number to get the next page.
+			options.PageNumber = ocList.NextPage
+		}
+		if len(ocMatches) == 0 {
+			return fmt.Errorf("No OAuthClients found matching the given parameters")
+		}
+		if len(ocMatches) > 1 {
+			return fmt.Errorf("Too many OAuthClients were found to match the given parameters. Please narrow your search.")
+		}
+		oc = ocMatches[0]
```

Comment:
Good idea! I'll make this change.

Reply:
Good idea. I've refactored this function out. I couldn't think of any better/shorter name so I used yours 🤣 .

(Agreed it would be nice to be able to search like this via the API directly.)